AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/DataStoreConfiguration.swift

```diff
@@ -63,16 +63,20 @@ public struct DataStoreConfiguration {
     /// The page size of each sync execution
     public let syncPageSize: UInt
 
+    public let syncExpressions: [DataStoreSyncExpression]
+
     init(errorHandler: @escaping DataStoreErrorHandler,
          conflictHandler: @escaping DataStoreConflictHandler,
          syncInterval: TimeInterval,
          syncMaxRecords: UInt,
-         syncPageSize: UInt) {
+         syncPageSize: UInt,
+         syncExpressions: [DataStoreSyncExpression]) {
         self.errorHandler = errorHandler
         self.conflictHandler = conflictHandler
         self.syncInterval = syncInterval
         self.syncMaxRecords = syncMaxRecords
         self.syncPageSize = syncPageSize
+        self.syncExpressions = syncExpressions
```

Comment:
@palpatim mentioned to combine a lot of these sync config parameters, but doing it this way seems to align better with android:
https://github.com/aws-amplify/amplify-android/blob/baa798bda6c97e9c8415792c4efcdc9dffb7ee3d/aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java#L60

Reply:
As long as our ultimate behavior is the same, I'm not too interested in hewing closely to Android platform conventions. Even with named arguments and defaults, a pattern like:

```swift
let foo = DataStoreConfiguration(
    errorHandler: self.handleDataStoreError,
    conflictHandler: self.handleDataStoreConflict,
    syncInterval: 86_400,
    syncMaxRecords: 10_000,
    syncPageSize: 100,
    syncExpressions: [(Post.foo == getLatestValue()), (Post.bar < getSomeOtherValue()) ]
)
```

is pretty unwieldy. Even if we provide defaults, the named arguments must be provided in the same order, which adds friction to the customer's implementation.

Consider encapsulating the initializer arguments in purpose-built types (we have `pluginOptions` on our config types), or adopting a different construction pattern (Builder, for example) to let the customer have an easier time of initializing these.

Reply:
> The overall design makes sense, but I'm not joyful about the growth of the DataStoreConfiguration interface, and I don't like passing that whole configuration object around to the underlying components. Comments inline

Regrading the configuraiton object being unwieldy, that is the worst case scenario.  For customers only wanting to specify a selective sync, it would look something like this:

```
            let syncExpression1 = DataStoreSyncExpression.syncExpression(Todo.schema, where: {
                Todo.keys.tag == AppDelegate.tagToSearch
            })
            let dataStorePlugin = AWSDataStorePlugin(modelRegistration: AmplifyModels(),
                                                     configuration: .custom(syncExpressions: [syncExpression1]))
```

Are you still concerned about bloating this interface?

Reply:
Error handler and conflict handler seem like reasonable things to specify as well, but I'm guessing we don't have any data on how many customers are using which of the config options. It seems to me that 4 free-floating and yet sync-related arguments is excessive, but it's ultimately a judgement call, so consider this as my vote against it. I'm willing to be outvoted. :)